core/src/chain/header.rs

```diff
@@ -0,0 +1,276 @@
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash as _Hash, Hasher};
+
+use chain::entry::Entry;
+use chain::SourceChain;
+
+// @TODO - serialize properties as defined in HeadersEntrySchema from golang alpha 1
+// @see https://github.com/holochain/holochain-proto/blob/4d1b8c8a926e79dfe8deaa7d759f930b66a5314f/entry_headers.go#L7
+// @see https://github.com/holochain/holochain-rust/issues/75
+#[derive(Clone, Debug, PartialEq, Serialize, Deserialize)]
+pub struct Header {
+    /// the type of this entry
+    /// system types may have associated "subconscious" behavior
+    entry_type: String,
+    /// ISO8601 time stamp
+    time: String,
+    /// link to the immediately preceding header, None is valid only for genesis
+    next: Option<u64>,
+    /// mandatory link to the entry for this header
+    entry: u64,
+    /// link to the most recent header of the same type, None is valid only for the first of type
+    type_next: Option<u64>,
+    /// agent's cryptographic signature
+    signature: String,
+}
+
+impl _Hash for Header {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.entry_type.hash(state);
+        self.time.hash(state);
+        self.next.hash(state);
+        self.entry.hash(state);
+        self.type_next.hash(state);
+        self.signature.hash(state);
+    }
+}
+
+impl Header {
+    /// build a new Header from a chain, entry type and entry.
+    /// a Header is immutable, but the chain is mutable if chain.push() is used.
+    /// this means that a header becomes invalid and useless as soon as the chain is mutated
+    /// the only valid usage of a header is to immediately push it onto a chain in a Pair.
+    /// normally (outside unit tests) the generation of valid headers is internal to the
+    /// chain::SourceChain trait and should not need to be handled manually
+    /// @see chain::pair::Pair
+    /// @see chain::entry::Entry
+    pub fn new<'de, C: SourceChain<'de>>(chain: &C, entry_type: &str, entry: &Entry) -> Header {
+        Header {
+            entry_type: entry_type.to_string(),
```

Comment:
that's a neat trick, i'd actually like to avoid even that as what i ultimately want is just:

`Header::new(&chain, &agent, &entry)`
`Pair::new(&chain, &agent, &entry)`
`Chain::push(&agent, &entry)`

which would mean putting things like time and type on the entry within rust, which would be copied into the `Header` for serialization and hashing for interop and protocol compatibility.

https://github.com/holochain/holochain-rust/issues/73